validator.go

```diff
@@ -0,0 +1,161 @@
+package validate
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"go.uber.org/multierr"
+)
+
+// FieldNameFunc is a function which converts a given reflect.StructField to a
+// string. The default will lookup json, yaml, and form field tags.
+type FieldNameFunc func(reflect.StructField) string
+
+// FieldJoinFunc joins a path slice with a given field. Both path and field may
+// be empty values.
+type FieldJoinFunc func(path []string, field string) string
+
+// Validator validates Validatable objects.
+type Validator struct {
+	fieldName FieldNameFunc
+	fieldJoin FieldJoinFunc
+}
+
+// New creates a new Validator.
+func New() *Validator {
+	return &Validator{}
+}
+
+// Validate will validate the given object. Structs, maps, slices, and arrays
+// will have each of their fields/items validated, effectively performing a
+// deep-validation.
+func (s *Validator) Validate(data interface{}) error {
+	if s.fieldName == nil {
+		s.fieldName = DefaultFieldName
+	}
+
+	if s.fieldJoin == nil {
+		s.fieldJoin = DefaultFieldJoin
+	}
+
+	return s.validate(nil, data)
+}
+
+// FieldNameFunc allows setting a custom FieldNameFunc method. It receives a
+// reflect.StructField, and must return a string for the name of that field. If
+// the returned string is empty, validation will not run against the field's
+// value, or any nested data within.
+func (s *Validator) FieldNameFunc(f FieldNameFunc) {
```

Comment:
My only suggestion is perhaps rather than letting this be overriden with a function here, you could let users provide a Method on their structs that fulfils some interface you check and use if it's implemented.

Reply:
That's definitely an idea I want to implement, but I'll leave that for a later PR for now :)

As for the existing ability to set a custom func applied to all structs, I think we should keep that too, as to prevent needing to add a custom func to all structs if you want to globally change it.